File: android/app/src/main/java/com/recre/app/core/calculo/Calculo.kt
package com.recre.app.core.calculo

import java.math.BigDecimal
import java.math.RoundingMode

/**
 * Single Source of Truth Kotlin del cálculo de recaudación.
 *
 * Espejo bit-a-bit de `supabase/functions/_shared/calculo.ts`:
 *
 *   1. créditos_netos = (Δentradas) − (Δsalidas)
 *   2. bruto = créditos × valor_credito  → setScale(2, HALF_UP)
 *   3. tasa_total = semanas × tasa_semanal  → setScale(2, HALF_UP)
 *   4. si bruto < tasa_total → procede=false (no se recauda)
 *   5. neto = bruto − tasa_total  (sin redondear; ya en céntimos)
 *   6. parte_local = neto × % / 100  → setScale(2, HALF_UP)
 *   7. parte_empresa = neto − parte_local  (la empresa absorbe el redondeo)
 *
 * El servidor recalcula todo igualmente al persistir (T-21) — esta
 * implementación se usa para:
 *   - Preview en vivo cuando el técnico introduce los contadores (T-54).
 *   - Validar suma de denominaciones contra `bruto` y `parte_local` (T-55).
 *   - Funcionar offline.
 *
 * Si los redondeos no coinciden bit-a-bit con el servidor, la fila quedará
 * marcada como conflicto en la base de datos (T-21 detecta deltas) y la
 * web ofrecerá la resolución manual (T-37).
 */
fun calcularRecaudacion(input: CalcularInput): Cifras {
    val valorCredito = input.valorCredito.setScale(SCALE, RoundingMode.HALF_UP)
    val tasaSemanal = input.tasaSemanal.setScale(SCALE, RoundingMode.HALF_UP)
    val porcentajeLocal = input.porcentajeLocal.setScale(SCALE, RoundingMode.HALF_UP)

    val deltaEntradas = input.contadorEntradasActual - input.baselineEntradas
    val deltaSalidas = input.contadorSalidasActual - input.baselineSalidas
    val creditos = deltaEntradas - deltaSalidas

    val brutoReal = valorCredito.multiply(BigDecimal(creditos))
        .setScale(SCALE, RoundingMode.HALF_UP)
    val tasaTotal = tasaSemanal.multiply(BigDecimal(input.semanas))
        .setScale(SCALE, RoundingMode.HALF_UP)

    // `procede` se decide con el bruto REAL: el dinero de verdad manda. El
    // redondeo solo cambia cómo se presenta una recaudación que ya procede.
    if (brutoReal < tasaTotal) {
        return Cifras(
            procede = false,
            bruto = brutoReal,
            semanas = input.semanas,
            tasaSemanal = tasaSemanal,
            tasaTotal = tasaTotal,
            neto = ZERO,
            porcentajeLocal = porcentajeLocal,
            parteLocal = ZERO,
            parteEmpresa = ZERO,
            valorCredito = valorCredito,
            baselineEntradas = input.baselineEntradas,
            baselineSalidas = input.baselineSalidas,
            deltaEntradas = deltaEntradas,
            deltaSalidas = deltaSalidas,
            creditos = creditos,
            redondeoAplicado = 0,
            brutoReal = brutoReal,
        )
    }

    // Redondeo opcional del bruto (config por empresa). Falsea la lectura de
    // salidas para que el bruto caiga en el múltiplo de `redondeoUnidad` más
    // cercano. El servidor persiste el contador ajustado y la diferencia se
    // arrastra a la siguiente recaudación; aquí solo redondeamos el bruto para
    // que el preview y el desglose cuadren con el resultado oficial.
    var bruto = brutoReal
    var redondeoAplicado = 0
    if (input.redondeoUnidad > 0) {
        val unidad = BigDecimal(input.redondeoUnidad)
        val ratio = brutoReal.divide(unidad, 0, RoundingMode.HALF_UP)
        var brutoObjetivo = ratio.multiply(unidad)
        // El redondeo nunca puede dejar el bruto por debajo de la tasa (neto < 0).
        if (brutoObjetivo < tasaTotal) {
            brutoObjetivo = brutoReal.divide(unidad, 0, RoundingMode.CEILING).multiply(unidad)
        }
        // creditosObjetivo es entero (el contador no admite fracciones); con
        // valorCredito divisor de la unidad el bruto cae exacto.
        val creditosObjetivo = brutoObjetivo.divide(valorCredito, 0, RoundingMode.HALF_UP)
        bruto = creditosObjetivo.multiply(valorCredito).setScale(SCALE, RoundingMode.HALF_UP)
        redondeoAplicado = input.redondeoUnidad
    }

    val neto = bruto.subtract(tasaTotal) // ya en céntimos
    val parteLocal = neto.multiply(porcentajeLocal)
        .divide(BigDecimal(100), SCALE, RoundingMode.HALF_UP)
    val parteEmpresa = neto.subtract(parteLocal) // absorbe el centavo

    return Cifras(
        procede = true,
        bruto = bruto,
        semanas = input.semanas,
        tasaSemanal = tasaSemanal,
        tasaTotal = tasaTotal,
        neto = neto,
        porcentajeLocal = porcentajeLocal,
        parteLocal = parteLocal,
        parteEmpresa = parteEmpresa,
        valorCredito = valorCredito,
        baselineEntradas = input.baselineEntradas,
        baselineSalidas = input.baselineSalidas,
        deltaEntradas = deltaEntradas,
        deltaSalidas = deltaSalidas,
        creditos = creditos,
        redondeoAplicado = redondeoAplicado,
        brutoReal = brutoReal,
    )
}

/** Entrada del cálculo. Mantiene la baseline + contadores actuales + parámetros de la instalación. */
data class CalcularInput(
    val baselineEntradas: Long,
    val baselineSalidas: Long,
    val contadorEntradasActual: Long,
    val contadorSalidasActual: Long,
    val valorCredito: BigDecimal,
    val tasaSemanal: BigDecimal,
    val porcentajeLocal: BigDecimal,
    val semanas: Int,
    /** Unidad de redondeo del bruto (config por empresa; 0 = sin redondeo). */
    val redondeoUnidad: Int = 0,
)

private const val SCALE = 2
private val ZERO: BigDecimal = BigDecimal.ZERO.setScale(SCALE, RoundingMode.UNNECESSARY)

File: supabase/functions/_shared/calculo.ts
/**
 * Single Source of Truth del cálculo de recaudación.
 *
 * Esta función se llama desde `calcular-recaudacion` (preview) y desde
 * `crear-recaudacion` (persistencia). Web y Android NUNCA recalculan
 * cifras: consumen el resultado de estas Edge Functions.
 *
 * Reglas (ver `.kiro/specs/recre/design.md §5`):
 *   1. créditos_netos = (Δentradas) − (Δsalidas)
 *   2. bruto = créditos_netos × valor_credito
 *   3. tasa_total = semanas_iso × tasa_semanal
 *   4. si bruto < tasa_total → no procede
 *   5. neto = bruto − tasa_total
 *   6. parte_local = round_half_up(neto × % / 100, 2)
 *   7. parte_empresa = neto − parte_local  (la empresa absorbe el redondeo)
 */

import { Decimal } from "decimal.js";

import type { BaselineInfo, CalculoRecaudacionResult, DenominacionItem } from "./types.ts";

// Configuración global de Decimal: redondeo HALF_UP, suficiente precisión.
Decimal.set({ rounding: Decimal.ROUND_HALF_UP, precision: 30 });

const CENTIMOS = 2;

export interface CalcularInput {
  baseline: BaselineInfo;
  contadorEntradasActual: number;
  contadorSalidasActual: number;
  /** Recibido como string para no perder precisión. */
  valorCredito: string;
  tasaSemanal: string;
  porcentajeLocal: string;
  semanas: number;
  /**
   * Unidad de redondeo del bruto (config por empresa; 0 = sin redondeo).
   * Si > 0, el bruto se lleva al múltiplo más cercano de esta unidad
   * falseando la lectura de salidas: ese contador ajustado se persiste como
   * real y la diferencia rueda a la siguiente recaudación vía baseline.
   */
  redondeoUnidad?: number;
}

/**
 * Calcula la recaudación a partir de la baseline y los contadores actuales.
 * Devuelve los importes serializados como string para preservar precisión
 * al cruzar la frontera HTTP.
 */
export function calcularRecaudacion(input: CalcularInput): CalculoRecaudacionResult {
  const valorCredito = new Decimal(input.valorCredito);
  const tasaSemanal = new Decimal(input.tasaSemanal);
  const porcentajeLocal = new Decimal(input.porcentajeLocal);

  const entradasDiff = new Decimal(input.contadorEntradasActual).minus(input.baseline.entradas);
  const salidasDiff = new Decimal(input.contadorSalidasActual).minus(input.baseline.salidas);
  const creditos = entradasDiff.minus(salidasDiff);

  const brutoReal = creditos.times(valorCredito).toDecimalPlaces(CENTIMOS);
  const tasaTotal = tasaSemanal.times(input.semanas).toDecimalPlaces(CENTIMOS);

  // `procede` se decide con el bruto REAL: el dinero de verdad manda. El
  // redondeo solo cambia cómo se presenta y guarda una recaudación que ya
  // procede; nunca convierte en recaudable lo que no llega a la tasa.
  if (brutoReal.lessThan(tasaTotal)) {
    return {
      procede: false,
      bruto: brutoReal.toFixed(CENTIMOS),
      semanas: input.semanas,
      tasa_semanal: tasaSemanal.toFixed(CENTIMOS),
      tasa_total: tasaTotal.toFixed(CENTIMOS),
      neto: "0.00",
      porcentaje_local: porcentajeLocal.toFixed(CENTIMOS),
      parte_local: "0.00",
      parte_empresa: "0.00",
      valor_credito: valorCredito.toFixed(CENTIMOS),
      baseline: input.baseline,
      contador_salidas_ajustado: input.contadorSalidasActual,
      recaudacion_bruta_real: brutoReal.toFixed(CENTIMOS),
      redondeo_aplicado: 0,
    };
  }

  // Redondeo opcional del bruto (config por empresa). Falseamos la lectura de
  // salidas lo justo para que el bruto caiga en el múltiplo de `redondeoUnidad`
  // más cercano; ese contador ajustado se persiste como real y la diferencia
  // rueda a la siguiente recaudación vía baseline (obtener_baseline). No se
  // pierde ni se inventa dinero: solo se reparte en cortes "redondos".
  const redondeoUnidad = input.redondeoUnidad ?? 0;
  let bruto = brutoReal;
  let salidasAjustado = input.contadorSalidasActual;
  let redondeoAplicado = 0;

  if (redondeoUnidad > 0) {
    const unidad = new Decimal(redondeoUnidad);
    const ratio = brutoReal.dividedBy(unidad);
    let brutoObjetivo = ratio.toDecimalPlaces(0).times(unidad);
    // El redondeo nunca puede dejar el bruto por debajo de la tasa (neto < 0):
    // en la rara franja justo encima de la tasa, redondeamos hacia arriba.
    if (brutoObjetivo.lessThan(tasaTotal)) {
      brutoObjetivo = ratio.toDecimalPlaces(0, Decimal.ROUND_CEIL).times(unidad);
    }
    // El contador es entero: creditosObjetivo se redondea al crédito más
    // cercano. Con `valorCredito` divisor de la unidad (p.ej. 0,20 y 10) el
    // bruto cae exacto; si no, queda en el múltiplo de valorCredito más
    // próximo al objetivo (diferencia < valorCredito, documentada en tests).
    const creditosObjetivo = brutoObjetivo.dividedBy(valorCredito).toDecimalPlaces(0);
    bruto = creditosObjetivo.times(valorCredito).toDecimalPlaces(CENTIMOS);
    // Créditos "de menos" respecto a lo real => subir salidas esa cantidad
    // (más salidas = menos créditos netos = menos bruto), y viceversa.
    const ajuste = creditos.minus(creditosObjetivo);
    salidasAjustado = new Decimal(input.contadorSalidasActual).plus(ajuste).toNumber();
    redondeoAplicado = redondeoUnidad;
  }

  const neto = bruto.minus(tasaTotal); // ya truncado a céntimos
  const parteLocal = neto.times(porcentajeLocal).dividedBy(100).toDecimalPlaces(CENTIMOS);
  const parteEmpresa = neto.minus(parteLocal); // absorbe el redondeo de céntimos

  return {
    procede: true,
    bruto: bruto.toFixed(CENTIMOS),
    semanas: input.semanas,
    tasa_semanal: tasaSemanal.toFixed(CENTIMOS),
    tasa_total: tasaTotal.toFixed(CENTIMOS),
    neto: neto.toFixed(CENTIMOS),
    porcentaje_local: porcentajeLocal.toFixed(CENTIMOS),
    parte_local: parteLocal.toFixed(CENTIMOS),
    parte_empresa: parteEmpresa.toFixed(CENTIMOS),
    valor_credito: valorCredito.toFixed(CENTIMOS),
    baseline: input.baseline,
    contador_salidas_ajustado: salidasAjustado,
    recaudacion_bruta_real: brutoReal.toFixed(CENTIMOS),
    redondeo_aplicado: redondeoAplicado,
  };
}

/**
 * Suma el valor económico de un desglose de denominaciones.
 * Usa Decimal para evitar errores de coma flotante.
 */
export function sumarDesglose(desglose: readonly DenominacionItem[]): string {
  let total = new Decimal(0);
  for (const item of desglose) {
    total = total.plus(new Decimal(item.denominacion).times(item.cantidad));
  }
  return total.toDecimalPlaces(CENTIMOS).toFixed(CENTIMOS);
}

/** Compara dos importes en string con tolerancia a representación. */
export function importesIguales(a: string, b: string): boolean {
  return new Decimal(a).equals(new Decimal(b));
}
